mitodl/ocw-to-hugo
# Refactor path handling

#### Pre-Flight checklist


#### What are the relevant tickets?
Related to #337 

#### What's this PR do?
This was part of #337 but I broke it out to be an easier review since this is mostly just a refactor with a couple exceptions. This does a few things:
 - Course feature links don't start with a `/` anymore. For example `/pages/projects` is now `pages/projects`. I think this is more intuitive as a relative link, it simplifies the cases we need to consider for creating links, and it should require only a tiny change in the template to handle it
 - The `rootRelativeToDocumentRelative` string replace code is removed. Instead we are making links with or without the slash prefix in `constructInternalLink`
 - `constructInternalLink` is meant to centralize the logic on what links should be rendered when, if shortcodes are allowed to be used or if the link should be relative to the course home page. #337 will expand this and add a resource shortcode which will be included as appropriate
 - PDF links are now created in `pathLookup`, with a new property `unalteredPath` which is used in some cases to resolve relative links
 - The whole path of a resource is matched, and in some cases the pdfs we export have different `technical_location` urls. I tried a few links from different courses where they differed and they still pointed to identical files so I don't know if the end user will see any improvement here. However you might notice that the uid of the course file is included in the `technical_location` now so I think it is probably resolving more correctly than before.

#### How should this be manually tested?
Convert all courses on master, then convert all courses on this PR, and compare the two directories. You probably won't be able to check everything but pick a few files which are different and verify that the changes you see make sense.

You should see some differences:
 - course feature links now start with `pages` instead of `/pages`
 - `technical_location` urls will be different in some cases
 - links to other courses in course descriptions now start with `/courses/...` instead of `courses/...` which should fix those links
 - `/ans7870` links in course descriptions are fixed and now start with a slash

I noticed a couple differences where the old behavior is broken so I don't consider these regressions:
 - A pdf link was updated and doesn't work, but it was also broken before so I am not counting this as a regression https://github.com/mitodl/ocw-to-hugo/issues/343
 - Another link was changed but that link was also broken beforehand #344
